File: architecture-skills/design-patterns/examples/proxy/proxy.py
"""Proxy pattern in Python (virtual proxy: lazy-loading image).

Duck typing means no formal interface is needed: the proxy just quacks
like the real image. __getattr__ forwards anything the proxy doesn't
override to the real subject, so only intercepted methods are written
by hand.
"""

from __future__ import annotations


class RealImage:
    """Real subject: expensive to create."""

    def __init__(self, filename: str) -> None:
        self.filename = filename
        print(f"Loading {filename} from disk (expensive)")

    def display(self) -> None:
        print(f"Displaying {self.filename}")

    def width(self) -> int:
        return 1920


class ImageProxy:
    """Virtual proxy: defers creating RealImage until first use."""

    def __init__(self, filename: str) -> None:
        self.filename = filename
        self._real: RealImage | None = None

    def display(self) -> None:
        if self._real is None:
            self._real = RealImage(self.filename)
        self._real.display()

    def __getattr__(self, name: str):
        # Forward anything else to the real subject, loading it on demand.
        if self._real is None:
            self._real = RealImage(self.filename)
        return getattr(self._real, name)


if __name__ == "__main__":
    gallery = [ImageProxy("holiday.png"), ImageProxy("unused.png")]
    print("Proxies created; nothing loaded yet")

    gallery[0].display()
    gallery[0].display()  # loaded once, displayed twice
    print(f"width via __getattr__: {gallery[0].width()}")

    print("unused.png was never loaded")

File: architecture-skills/design-patterns/examples/proxy/proxy.swift
// Proxy pattern in Swift (virtual proxy: lazy-loading image).
//
// A protocol is the subject interface; the proxy is a class holding an
// optional real subject created on first use. A stored `lazy var`
// covers the simpler case where property access alone should trigger
// creation.

protocol Image {
    func display()
}

/// Real subject: expensive to create.
final class RealImage: Image {
    private let filename: String

    init(filename: String) {
        self.filename = filename
        print("Loading \(filename) from disk (expensive)")
    }

    func display() {
        print("Displaying \(filename)")
    }
}

/// Virtual proxy: defers creating RealImage until first use.
final class ImageProxy: Image {
    private let filename: String
    private var real: RealImage?

    init(filename: String) {
        self.filename = filename
    }

    func display() {
        if real == nil {
            real = RealImage(filename: filename)
        }
        real?.display()
    }
}

let gallery: [Image] = [
    ImageProxy(filename: "holiday.png"),
    ImageProxy(filename: "unused.png"),
]
print("Proxies created; nothing loaded yet")

gallery[0].display()
gallery[0].display() // loaded once, displayed twice

print("unused.png was never loaded")
